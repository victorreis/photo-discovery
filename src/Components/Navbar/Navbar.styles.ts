import styled from 'styled-components';

import { ScreenBreakpoint, screenBreakpoints } from '../../Theme/Types';
import { Typography } from '../Typography';

export const heightByScreenSize: Record<
  Extract<ScreenBreakpoint, 'xs'> | 'others',
  string
> = {
  others: '5rem',
  xs: '8rem',
};

export const NavbarContainer = styled.div`
  padding: 0 2rem;
  height: ${heightByScreenSize.others};
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  flex: 1;

  @media only screen and (max-width: ${screenBreakpoints.xs}px) {
    height: ${heightByScreenSize.xs};
  }
`;

export const Logo = styled(Typography)`
  text-transform: uppercase;
  justify-content: center;

  ${({ theme }) =>
    `background: linear-gradient(135deg,
     ${theme.colors.main.effect.dark},
     ${theme.colors.main.effect.light},
     ${theme.colors.main.detail.darker},
     ${theme.colors.main.detail.normal});`}
  background-clip: text;
  -webkit-background-clip: text;
  -webkit-text-fill-color: transparent;

  span.larger {
    font-size: 2.2rem;
  }

  @media only screen and (max-width: ${screenBreakpoints.xs}px) {
    display: flex;
    flex-direction: column;
    height: ${heightByScreenSize.xs};
  }
`;
