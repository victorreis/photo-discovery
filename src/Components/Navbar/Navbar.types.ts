import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredNavbarProps {}

export interface DefaultNavbarProps {}

export interface OptionalNavbarProps {}

export type NavbarProps = RequiredNavbarProps &
  DefaultNavbarProps &
  OptionalNavbarProps &
  TestProps;

export type NavbarStyleProps = Required<DefaultNavbarProps>;
