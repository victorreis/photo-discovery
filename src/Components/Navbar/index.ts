export * from './Navbar.types';
export * from './Navbar';
