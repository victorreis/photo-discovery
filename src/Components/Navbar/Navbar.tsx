import { HiColorSwatch } from 'react-icons/hi';

import { useTheme } from 'styled-components';

import { TestProps } from '../../Config/Tests/Test.types';
import { Logo, NavbarContainer } from './Navbar.styles';
import { DefaultNavbarProps, NavbarProps } from './Navbar.types';

export const navbarDefaults: Required<DefaultNavbarProps> &
  Required<TestProps> = {
  testID: 'Navbar',
};

export const Navbar: React.FC<NavbarProps> = (props): JSX.Element => {
  const theme = useTheme();
  const { testID = navbarDefaults.testID } = props;

  return (
    <NavbarContainer data-testid={testID}>
      <Logo variant="h3">
        <span>
          <span className="larger">P</span>
          hoto
        </span>

        <span>
          <span className="larger">D</span>
          iscovery
        </span>
      </Logo>

      <HiColorSwatch fill={theme.colors.main.effect.light} size={24} />
    </NavbarContainer>
  );
};
