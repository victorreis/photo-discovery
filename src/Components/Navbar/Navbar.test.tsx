import {
  fireEvent,
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { modalDefaults } from '../Modal/Modal';
import { Navbar, navbarDefaults } from './Navbar';
import { heightByScreenSize } from './Navbar.styles';
import { RequiredNavbarProps, NavbarProps } from './Navbar.types';

describe('navbar component tests', () => {
  const text = 'text';
  const themeSwitcherIconId = `${navbarDefaults.testID}_ThemeSwitcherIcon`;
  const modalBackdropId = `${modalDefaults.testID}_ModalBackdropContainer`;

  const requiredProps: RequiredNavbarProps = {
    children: text,
  };

  const setup = (props?: NavbarProps) => {
    const renderRTR = () =>
      renderRTRCreator<NavbarProps>(Navbar, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<NavbarProps>(Navbar, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(navbarDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should open the modal when the '${themeSwitcherIconId}' is clicked`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const themeSwitcherIcon = screen.getByTestId(themeSwitcherIconId);

      fireEvent.click(themeSwitcherIcon);
      const modal = screen.getByTestId(modalDefaults.testID);

      expect(modal).toBeInTheDocument();
    });

    it(`should close the modal when the '${themeSwitcherIconId}' is clicked and then the '${modalBackdropId}' is clicked`, () => {
      expect.assertions(2);
      setup().renderJestDom();
      const themeSwitcherIcon = screen.getByTestId(themeSwitcherIconId);

      fireEvent.click(themeSwitcherIcon);
      const modal = screen.getByTestId(modalDefaults.testID);
      const modalBackdrop = screen.getByTestId(modalBackdropId);

      expect(modal).toBeInTheDocument();

      fireEvent.click(modalBackdrop);

      expect(modal).not.toBeInTheDocument();
    });
  });

  describe('style tests', () => {
    it(`should have height equals '${heightByScreenSize.others}' when the screen is large`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(navbarDefaults.testID);

      expect(container).toHaveStyle({
        height: heightByScreenSize.others,
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
