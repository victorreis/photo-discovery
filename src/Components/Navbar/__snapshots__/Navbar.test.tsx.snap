// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar component tests snapshot tests should render correctly 1`] = `
<div
  className="NavbarContainer-photo-discovery__sc-9em23l-0 lmUnAu"
  data-testid="Navbar"
>
  <h3
    className="TypographyContainer-photo-discovery__sc-1t3hmm1-0 gHcSmf Logo-photo-discovery__sc-9em23l-1 cSizAu"
    data-testid="Typography"
  >
    <span>
      <span
        className="larger"
      >
        P
      </span>
      hoto
    </span>
    <span>
      <span
        className="larger"
      >
        D
      </span>
      iscovery
    </span>
  </h3>
  <svg
    data-testid="Navbar_ThemeSwitcherIcon"
    fill="#f480ba"
    height="1.67rem"
    onClick={[Function]}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 20 20"
    width="1.67rem"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M4 2a2 2 0 00-2 2v11a3 3 0 106 0V4a2 2 0 00-2-2H4zm1 14a1 1 0 100-2 1 1 0 000 2zm5-1.757l4.9-4.9a2 2 0 000-2.828L13.485 5.1a2 2 0 00-2.828 0L10 5.757v8.486zM16 18H9.071l6-6H16a2 2 0 012 2v2a2 2 0 01-2 2z"
      fillRule="evenodd"
    />
  </svg>
</div>
`;
